### PR TITLE
Reconcile the dashboard library aggregator

### PR DESCRIPTION
**Pull request by rbarry82**
Bring it in sync with changes from the proxy.

Make it more resilient to charm lifecycle events, basically.

### CODE REVIEW

**Review by balbirthomas — APPROVED** [review]

**Comment by balbirthomas** [comment]
This looks like  it needs fixing
```
/helpers.py", line 171, in get_dashboard_by_search
    dashboard_json = await grafana.fetch_dashboard(dashboards[0]["uid"])
IndexError: list index out of range
```

**Comment by rbarry82** [comment]
Not related to this PR, but I'll fix the integration test, too. Possibly a timing-related failure.

**Review by sed-i — APPROVED** [review]

File: lib/charms/grafana_k8s/v0/grafana_dashboard.py [hunk 1]
```diff
@@ -1,7 +1,166 @@
...
+An instantiated `GrafanaDashboardProvider` validates that the path specified
+in the constructor (or the default) exists, reads the file contents, then
+compresses them with LZMA and adds them to the application relation data
+when a relation is established with Grafana.
+
+Provided dashboards will be checked by Grafana, and a series of dropdown menus
+providing the ability to select query targets by Juju Model, application instance,
+and unit will be added if they do not exist.
+
+To avoid requiring `jinja` in `GrafanaDashboardProvider` users, templte validation
```
Comment (balbirthomas) [inline lib/charms/grafana_k8s/v0/grafana_dashboard.py:67]: Typo: "templte"

File: lib/charms/grafana_k8s/v0/grafana_source.py [hunk 2]
```diff
@@ -1,7 +1,122 @@
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-"""A library for working with Grafana datasources for charm authors."""
+"""## Overview.
+
+This document explains how to integrate with the Grafana charm
+for the purpose of providing a datasource which can be used by
+Grafana dashboards. It also explains the structure of the data
+expected by the `grafana-dashboard` interface, and may provide a
```
Comment (balbirthomas) [inline lib/charms/grafana_k8s/v0/grafana_source.py:9]: Copy paste error ? : "expected by the `grafana-dashboard` ... . Should it be `grafana-source` ?

**Review by balbirthomas — APPROVED** [review]

**Review by balbirthomas — DISMISSED** [review]

**Comment by balbirthomas** [comment]
Oooh! MyPy Fun -
```
lib/charms/grafana_k8s/v0/grafana_source.py:353: error: Argument 2 to
"_validate_relation_by_interface_and_direction" has incompatible type
"Optional[str]"; expected "str"  [arg-type]
                charm, relation_name, RELATION_INTERFACE_NAME, RelationRol...
```

**Review by sed-i — APPROVED** [review]

**Review by sed-i — APPROVED** [review]

**Review by balbirthomas — APPROVED** [review]

**Pull request merged**